libnmstate/nm/dns.py

```diff
@@ -0,0 +1,101 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+from operator import itemgetter
+from operator import methodcaller
+
+from libnmstate import iplib
+from libnmstate.nm import nmclient
+from libnmstate.schema import DNS
+from libnmstate.nm import ipv4 as nm_ipv4
+from libnmstate.nm import ipv6 as nm_ipv6
+from libnmstate.error import NmstateInternalError
+
+
+DNS_DEFAULT_PRIORITY_VPN = 50
+DNS_DEFAULT_PRIORITY_OTHER = 100
+
+IPV6_ADDRESS_LENGTH = 128
+
+
+def get_running():
+    dns_state = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    client = nmclient.client()
+    dns_confs = client.get_dns_configuration()
+    for dns_conf in sorted(dns_confs, key=methodcaller('get_priority')):
+        iface_name = dns_conf.get_interface()
+        for ns in dns_conf.get_nameservers():
+            if iplib.is_ipv6_link_local_addr(ns, IPV6_ADDRESS_LENGTH):
+                if not iface_name:
+                    # For IPv6 link local address, the interface name should be
+                    # appended also.
+                    raise NmstateInternalError(
+                        "Got empty interface name for IPv6 link local "
+                        "DNS nameserver entry {}".format(ns))
+                ns_addr = "{}%{}".format(ns, iface_name)
+            else:
+                ns_addr = ns
+            dns_state[DNS.SERVER].append(ns_addr)
+        dns_state[DNS.SEARCH].extend(dns_conf.get_domains())
+    if not dns_state[DNS.SERVER] and not dns_state[DNS.SEARCH]:
+        return {}
+    return dns_state
+
+
+def get_config():
+    dns_conf = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    tmp_dns_confs = []
+    client = nmclient.client()
+    for ac in client.get_active_connections():
+        # NM prefers IPv6 over IPv4 DNS.
+        for ip_profile in (nm_ipv6.get_ip_profile(ac),
+                           nm_ipv4.get_ip_profile(ac)):
+            if not ip_profile:
+                continue
+            if not ip_profile.props.dns and not ip_profile.props.dns_search:
+                continue
+            priority = ip_profile.props.dns_priority
+            if priority == 0:
+                if ac.get_vpn():
+                    priority = DNS_DEFAULT_PRIORITY_VPN
+                else:
+                    priority = DNS_DEFAULT_PRIORITY_OTHER
+
+            tmp_dns_confs.append(
+                {
+                    'server': ip_profile.props.dns,
+                    'priority': priority,
+                    'search': ip_profile.props.dns_search,
+                }
+            )
+    # NetworkManager sorts the DNS entries based on whether the active
```

Comment:
How about adding a link to  https://developer.gnome.org/NetworkManager/stable/nm-settings.html#nm-settings.property.ipv4.dns-priority 

new text:

```
NetworkManager sorts the DNS entries based on various criteria including which profile was activated first when profiles are activated. Therefore the configuration does not completely define the order. To define the order in a declarative way, Nmstate only uses the priority to order the entries.
Reference:
https://developer.gnome.org/NetworkManager/stable/nm-settings.html#nm-settings.property.ipv4.dns-priority
```

I wonder if it makes sense to add this together with the code to actually set the DNS entries to give more context.

Reply:
In stead of using the long url, I used this instead:

```
    # NetworkManager sorts the DNS entries based on various criteria including
    # which profile was activated first when profiles are activated. Therefore
    # the configuration does not completely define the order. To define the
    # order in a declarative way, Nmstate only uses the priority to order the
    # entries. Please check NetworkManager document about ipv4.dns-priority for
    # more detail.
```

Reply:
IMHO a URL is a lot nicer, even when the URL is long.